File: linked_list/reversed_ll_recursion.cpp
/* Full implementation of the reversed linked list data structure using recursion in C++ */
#include <iostream>

/* Node class */
class Node
{
public:
    int m_data;
    Node *next;
    // Default constructor
    Node() = default;
    // Parameterized constructor
    Node(int data)
        : m_data(data), next(NULL)
    {
    }
};

/* Linked list class */
class LinkedList
{
public:
    Node *head;

public:
    // Default constructor
    LinkedList()
        : head(NULL)
    {
    }
    /* Public member functions */

    // Push a new node to the linked list
    void pushNode(int data)
    {
        // Dynamic memory allocation on the heap for a new node
        Node *newNode = new Node();
        // Assign values
        newNode->m_data = data;
        // Point next to the head
        newNode->next = head;
        // Shift the head to the new node
        head = newNode;
    }

    // Reverse the linked list using recurson
    Node *reverseList(Node *head)
    {
        // Let's assume there's no head
        if (!head || !head->next)
        {
            return head;
        }
        // Divide the linked list in two parts: the first node and the rest, then reverse the rest
        Node *rest = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        // Fist the head pointer
        return rest;
    }

    // Print the linked list
    void printList()
    {
        // Set a temp variable to travese the list starting at the head
        Node *temp = head;
        while (temp)
        {
            std::cout << temp->m_data << " ";
            temp = temp->next;
        }
        std::cout << std::endl;
    }
};

int main()
{
    // Instantiate the Linked list class
    LinkedList ll;
    ll.pushNode(25);
    ll.pushNode(35);
    ll.pushNode(45);
    ll.pushNode(3);
    ll.pushNode(2);
    ll.printList();
    ll.head = ll.reverseList(ll.head);
    ll.printList();
    return 0;
}

File: linked_list/reversed_ll_recursion.py
# Full implementation of the reversed linked list DSA in Python using recursion

# Node class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Linked list class


class LinkedList:
    def __init__(self):
        self.head = None

    # Public methods

    # Push a new node to the linked list
    def push_node(self, data):
        # Create a new node
        new_node = Node(data)
        # Point the next to the head
        new_node.next = self.head
        # Point the head to the new node
        self.head = new_node

    # Reverse linked list
    def reverse_list(self, head):
        # Check if there's no head or next node
        if (head is None or head.next is None):
            return head
        # Divide the linked list in two parts: the first node and the rest of the list, then recursively reverse the rest of the list for the rest
        rest = self.reverse_list(head.next)
        # Put the first element at the end
        head.next.next = head
        head.next = None
        # Fix the head pointer
        return rest

    # Print linked list
    def print_list(self):
        # Traverse the linked list from the head and print values of every node on each iteration
        temp = self.head
        while (temp):
            print(f'Linked list: {temp.data} ', end='')
            temp = temp.next
        print()


if __name__ == '__main__':
    # Instantiate the ll class and make a new object of that class
    ll = LinkedList()
    ll.push_node(2)
    ll.push_node(5)
    ll.push_node(11)
    ll.print_list()
    ll.head = ll.reverse_list(ll.head)
    ll.print_list()
